### DISCUSSION TITLE
YATT (Yet Another TS0601 Thermostat) _TZE200_ye5jkfsb

### DISCUSSION THREAD

**Discussion by Westcott1** [Koenkk/zigbee2mqtt#9024] [Devices]
modelID: 'TS0601' 
manufacturerName: '_TZE200_ye5jkfsb'
model: 'ME81AH',
[https://www.aliexpress.com/item/1005002960679997.html?spm=a2g0s.9042311.0.0.7ac84c4dkawYw2](url)
Unsupported, so I'm trying to get an external converter to work.
Using `fromZigbee: [fz.tuya_thermostat]`
OK apart from local temperature 10x too low.
It seems this device reports the actual integer value.

**Comment by tboy72** [comment]
Hello Westscott1!
You could share your external converter with me, I have the same thermostat, but it doesn't work with z2m.

**Reply by Westcott1** [reply]
Sorry to be so slow in responding.
I still haven't got it to work correctly, the local_temperature is 10x too low.
I know how to fix it, but haven't done so yet.
However, here is my converter -

const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require("zigbee-herdsman-converters/lib/tuya");

module.exports = [
    {
        fingerprint: [{modelID: 'TS0601', manufacturerName: '_TZE200_ye5jkfsb'},],
        model: 'ME81AH',
        vendor: 'TuYa',
        description: 'Wall-mount thermostat',
                fromZigbee: [fz.moes_thermostat],
        toZigbee: [tz.moes_thermostat_child_lock, tz.moes_thermostat_current_heating_setpoint, tz.moes_thermostat_mode,
            tz.moes_thermostat_standby, tz.moes_thermostat_sensor, tz.moes_thermostat_calibration,
            tz.moes_thermostat_deadzone_temperature, tz.moes_thermostat_max_temperature_limit],
        exposes: [e.child_lock(), e.deadzone_temperature(), e.max_temperature_limit(),
            exposes.climate().withSetpoint('current_heating_setpoint', 5, 30, 1, ea.STATE_SET)
                .withLocalTemperature(ea.STATE).withLocalTemperatureCalibration(ea.STATE_SET)
                .withSystemMode(['off', 'heat'], ea.STATE_SET).withRunningState(['idle', 'heat', 'cool'], ea.STATE)
                .withPreset(['hold', 'program']).withSensor(['IN', 'AL', 'OU'], ea.STATE_SET)],
        onEvent: tuya.onEventSetLocalTime,
    },
];

**Reply by tboy72** [reply]
Thank you very much, you helped a lot!

**Reply by tboy72** [reply]
If you manage to resolve the bad placeholder display, will you share again? Thx!

**Reply by revnyolc** [reply]
:(
can not read a block mapping entry; a multiline key may not be an implicit key at line 14, column 12:
    fingerprint: [{modelID: 'TS0601', manufactu ... 
               ^

**Reply by Westcott1** [reply]
Most likely a syntax error in your file.
Can you post your file by dragging and dropping, rather than pasting?

**Reply by revnyolc** [reply]
I copied your code one by one.
I added the /config/zigbee2mqtt/configuration.yaml file like this:

```
  "external_converters": [
      "thermostat.js"],
  "devices": [
    "devices.yaml"
  ],
```
**( After restarting Zigbee2MQTT, this add-on will disappear from the file! :( )**

Then I created the /config/zigbee2mqtt/termosztat.js file with your code:

```
const fz = require ('zigbee-herdsman-converters / converters / fromZigbee');
const tz = require ('zigbee-herdsman-converters / converters / toZigbee');
const exposes = require ('zigbee-herdsman-converters / lib / exposes');
const reporting = require ('zigbee-herdsman-converters / lib / reporting');
const extend = require ('zigbee-herdsman-converters / lib / extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require ("zigbee-herdsman-converters / lib / tuya");

module.exports = [
{
fingerprint: [{modelID: 'TS0601', manufacturerName: '_TZE200_ye5jkfsb', friendly_name: 'thermostat_name'},],
model: 'ME81AH',
vendor: 'TuYa',
description: 'Wall-mount thermostat',
fromZigbee: [fz.moes_thermostat],
toZigbee: [tz.moes_thermostat_child_lock, tz.moes_thermostat_current_heating_setpoint, tz.moes_thermostat_mode,
tz.moes_thermostat_standby, tz.moes_thermostat_sensor, tz.moes_thermostat_calibration,
tz.moes_thermostat_deadzone_temperature, tz.moes_thermostat_max_temperature_limit],
exposes: [e.child_lock (), e.deadzone_temperature (), e.max_temperature_limit (),
exposes.climate (). withSetpoint ('current_heating_setpoint', 5, 30, 1, ea.STATE_SET)
.withLocalTemperature (ea.STATE) .withLocalTemperatureCalibration (ea.STATE_SET)
.withSystemMode (['off', 'heat'], ea.STATE_SET) .withRunningState (['idle', 'heat', 'cool'], ea.STATE)
.withPreset (['hold', 'program']). withSensor (['IN', 'AL', 'OU'], ea.STATE_SET)],
onEvent: tuya.onEventSetLocalTime,
},
];
```

**Reply by revnyolc** [reply]
It works !!!!
The Z2M was old, it works with the new one!
https://github.com/zigbee2mqtt/hassio-zigbee2mqtt

Thank you very much for your help !!!

Please do it to share with others!
eg here:
https://githubmemory.com/repo/Koenkk/zigbee2mqtt/issues/9189

**Reply by Westcott1** [reply]
Good result.

**Comment by Fnordcorps** [comment]
I managed to get this working with my me81ah with your code. Works great, Thankyou! 

Did you ever solve the reported temperature being 10 x too low? I also have this problem. Otherwise works great

**Comment by tolwi** [comment]
Try this one:

```javaScript
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const e = exposes.presets;
const ea = exposes.access;

const fzDataPoints = {
    avattoSwitch: 1,
    avattoMode: 2,
    avattoTempSet: 16,
    avattoTempCurrent: 24,
    avattoChildLock: 40,
    avattoTempCorrection: 27,
    avattoSensor: 43,
};

const tzConverters = {
    avatto_thermostat_child_lock: {
        key: ['child_lock'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointBool(entity, fzDataPoints.avattoChildLock, value === 'lock');
        },
    },
    avatto_thermostat_preset: {
        key: ['preset'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointEnum(entity, fzDataPoints.avattoMode, value === 'manual' ? 0 : 1);
        },
    },
    avatto_thermostat_sensor: {
        key: ['sensor'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointEnum(entity, fzDataPoints.avattoSensor, value === 'in' ? 0 : 1);
        },
    },
    avatto_thermostat_mode: {
        key: ['system_mode'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointEnum(entity, fzDataPoints.avattoSwitch, value === 'heat');
        },
    },

    avatto_thermostat_temp_correction: {
        key: ['local_temperature_calibration'],
        convertSet: async (entity, key, value, meta) => {
            const temp = Math.round(value);
            await tuya.sendDataPointValue(entity, fzDataPoints.avattoTempCorrection, temp);
        },
    },
    avatto_thermostat_temp_set: {
        key: ['current_heating_setpoint'],
        convertSet: async (entity, key, value, meta) => {
            const temp = Math.round(value);
            await tuya.sendDataPointValue(entity, fzDataPoints.avattoTempSet, temp);
        },
    },

};

const fzConverters = {
    avatto_floor_thermostat: {
        cluster: 'manuSpecificTuya',
        type: ['commandGetData', 'commandSetDataResponse'],
        convert: (model, msg, publish, options, meta) => {
            const dp = msg.data.dp;
            const value = tuya.getDataValue(msg.data.datatype, msg.data.data);
            switch (dp) {
            case fzDataPoints.avattoSwitch:
                return {system_mode: value ? 'heat' : 'off'};
            case fzDataPoints.avattoChildLock: // mode
                return {child_lock: value ? 'lock' : 'unlock'};
            case fzDataPoints.avattoTempSet:
                return {current_heating_setpoint: value};
            case fzDataPoints.avattoTempCurrent:
                return {local_temperature: value};
            case fzDataPoints.avattoTempCorrection:
                return {local_temperature_calibration: value};
            case fzDataPoints.avattoMode:
                return {preset: value === 0 ? 'manual' : 'program'};
            case fzDataPoints.avattoSensor:
                return {sensor: value === 0 ? 'in' : 'out'};
            default:
                meta.logger.warn(`zigbee-herdsman-converters:AVATTO: Unrecognized DP #${dp} with data ${JSON.stringify(msg.data)}`);
            }
        },
    },
};

module.exports = [
    {
        fingerprint: [{modelID: 'TS0601', manufacturerName: '_TZE200_ye5jkfsb'}],
        model: 'ME81AH',
        vendor: 'TuYa',
        description: 'Wall-mount thermostat',
        whiteLabel: [
            {vendor: 'AVATTO', model: 'ME81AH'},
        ],
        fromZigbee: [fzConverters.avatto_floor_thermostat],
        toZigbee: [
            tzConverters.avatto_thermostat_child_lock,
            tzConverters.avatto_thermostat_preset,
            tzConverters.avatto_thermostat_mode,
            tzConverters.avatto_thermostat_temp_set,
            tzConverters.avatto_thermostat_temp_correction,
            tzConverters.avatto_thermostat_sensor,
        ],
        exposes: [
            e.child_lock(),
            e.local_temperature(),
            exposes.climate()
                .withSetpoint('current_heating_setpoint', 5, 30, 1, ea.STATE_SET)
                .withLocalTemperatureCalibration(ea.STATE_SET)
                .withSystemMode(['off', 'heat'])
                .withPreset(['manual', 'program'])
                .withSensor(['in', 'out'], ea.STATE_SET),
        ],
        onEvent: tuya.onEventSetLocalTime,
    },
];

```

**Reply by nussknacker4711** [reply]
Hi all, what a great code, thanks for this!
I have got 3 "Alelke Zigbee Temperature Controller", same fingerprint - and they work fine. I have connected them to fhem.  

Is it possible to read the heating state on/off anywhere? The system_mode seems to be the main switch (display and controller state), but not if it is currently heating or not.

When i change the temperature I got a lot of readings e.g.
Unrecognized DP #36 with data {"status":0,"transid":6,"dp":36,"datatype":4,"fn":0,"data":{"type":"Buffer","data":[0]}}
But not corresponding to the state.

best regards

Mike

**Reply by mifrith** [reply]
Hey! DP # 36 is the current state of the heater (idle / heating). Try this version of the converter:

```
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const e = exposes.presets;
const ea = exposes.access;

const fzDataPoints = {
    avattoSwitch: 1,
    avattoMode: 2,
    avattoTempSet: 16,
    avattoTempCurrent: 24,
    avattoChildLock: 40,
    avattoTempCorrection: 27,
    avattoSensor: 43,
    avattoRunningState: 36,
};

const tzConverters = {
    avatto_thermostat_child_lock: {
        key: ['child_lock'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointBool(entity, fzDataPoints.avattoChildLock, value === 'lock');
        },
    },
    avatto_thermostat_preset: {
        key: ['preset'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointEnum(entity, fzDataPoints.avattoMode, value === 'manual' ? 0 : 1);
        },
    },
    avatto_thermostat_sensor: {
        key: ['sensor'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointEnum(entity, fzDataPoints.avattoSensor, value === 'in' ? 0 : 1);
        },
    },
    avatto_thermostat_mode: {
        key: ['system_mode'],
        convertSet: async (entity, key, value, meta) => {
            await tuya.sendDataPointEnum(entity, fzDataPoints.avattoSwitch, value === 'heat');
        },
    },

    avatto_thermostat_temp_correction: {
        key: ['local_temperature_calibration'],
        convertSet: async (entity, key, value, meta) => {
            const temp = Math.round(value);
            await tuya.sendDataPointValue(entity, fzDataPoints.avattoTempCorrection, temp);
        },
    },
    avatto_thermostat_temp_set: {
        key: ['current_heating_setpoint'],
        convertSet: async (entity, key, value, meta) => {
            const temp = Math.round(value);
            await tuya.sendDataPointValue(entity, fzDataPoints.avattoTempSet, temp);
        },
    },

};

const fzConverters = {
    avatto_floor_thermostat: {
        cluster: 'manuSpecificTuya',
        type: ['commandGetData', 'commandSetDataResponse'],
        convert: (model, msg, publish, options, meta) => {
            const dp = msg.data.dp;
            const value = tuya.getDataValue(msg.data.datatype, msg.data.data);
            switch (dp) {
            case fzDataPoints.avattoSwitch:
                return {system_mode: value ? 'heat' : 'off'};
            case fzDataPoints.avattoChildLock:
                return {child_lock: value ? 'lock' : 'unlock'};
            case fzDataPoints.avattoTempSet:
                return {current_heating_setpoint: value};
            case fzDataPoints.avattoTempCurrent:
                return {local_temperature: value};
            case fzDataPoints.avattoTempCorrection:
                return {local_temperature_calibration: value};
            case fzDataPoints.avattoMode:
                return {preset: value === 0 ? 'manual' : 'program'};
            case fzDataPoints.avattoSensor:
                return {sensor: value === 0 ? 'in' : 'out'};
			case fzDataPoints.avattoRunningState:
                return {running_state: value === 0 ? 'heat' : 'idle'};
            default:
                meta.logger.warn(`zigbee-herdsman-converters:AVATTO: Unrecognized DP #${dp} with data ${JSON.stringify(msg.data)}`);
            }
        },
    },
};

module.exports = [
    {
        fingerprint: [{modelID: 'TS0601', manufacturerName: '_TZE200_ye5jkfsb'}],
        model: 'ME81AH',
        vendor: 'TuYa',
        description: 'Wall-mount thermostat',
        whiteLabel: [
            {vendor: 'AVATTO', model: 'ME81AH'},
        ],
        fromZigbee: [fzConverters.avatto_floor_thermostat],
        toZigbee: [
            tzConverters.avatto_thermostat_child_lock,
            tzConverters.avatto_thermostat_preset,
            tzConverters.avatto_thermostat_mode,
            tzConverters.avatto_thermostat_temp_set,
            tzConverters.avatto_thermostat_temp_correction,
            tzConverters.avatto_thermostat_sensor,
        ],
        exposes: [
            e.child_lock(),
            e.local_temperature(),
            exposes.climate()
                .withSetpoint('current_heating_setpoint', 5, 30, 1, ea.STATE_SET)
                .withLocalTemperatureCalibration(ea.STATE_SET)
                .withSystemMode(['off', 'heat'])
		.withRunningState(['idle', 'heat'],ea.STATE)
                .withPreset(['manual', 'program'])
                .withSensor(['in', 'out'], ea.STATE_SET),
        ],
        onEvent: tuya.onEventSetLocalTime,
    },
];

```
Thanks tolwi!

**Reply by nussknacker4711** [reply]
Hi mifrith 

thanks for this: The thermostat is quite lazy, between switching and sending its status it takes about 10 seconds, and: it works!

Great work

**Reply by tmaniac** [reply]
Thanks for the converter-config;

If i try to use this external converter i get a startup error in my HomeAssistant;

> Zigbee2MQTT:error 2021-12-09 10:02:36: Failed to call 'HomeAssistant' 'start' (AssertionError [ERR_ASSERTION]: No temperature found
>     at HomeAssistant.exposeToConfig (/app/lib/extension/homeassistant.ts:182:19)
>     at HomeAssistant.populateMapping (/app/lib/extension/homeassistant.ts:784:54)
>     at HomeAssistant.start (/app/lib/extension/homeassistant.ts:63:14)
>     at Controller.callExtensions (/app/lib/controller.ts:310:17)
>     at Controller.start (/app/lib/controller.ts:152:9)
>     at start (/app/index.js:97:5))

Is there something i'm doing wrong?

**Reply by mifrith** [reply]
Yes, there is a problem. But there is also a solution in the form of a crutch. When adding a new device, you use this converter:
```
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const e = exposes.presets;
const ea = exposes.access;
const tuya = require("zigbee-herdsman-converters/lib/tuya");

module.exports = [
{
fingerprint: [{modelID: 'TS0601', manufacturerName: '_TZE200_ye5jkfsb'},],
model: 'ME81AH',
vendor: 'TuYa',
description: 'Wall-mount thermostat',
fromZigbee: [fz.moes_thermostat],
toZigbee: [tz.moes_thermostat_child_lock, tz.moes_thermostat_current_heating_setpoint, tz.moes_thermostat_mode,
tz.moes_thermostat_standby, tz.moes_thermostat_sensor, tz.moes_thermostat_calibration,
tz.moes_thermostat_deadzone_temperature, tz.moes_thermostat_max_temperature_limit],
exposes: [e.child_lock(), e.deadzone_temperature(), e.max_temperature_limit(),
exposes.climate().withSetpoint('current_heating_setpoint', 5, 30, 1, ea.STATE_SET)
.withLocalTemperature(ea.STATE).withLocalTemperatureCalibration(ea.STATE_SET)
.withSystemMode(['off', 'heat'], ea.STATE_SET).withRunningState(['idle', 'heat', 'cool'], ea.STATE)
.withPreset(['hold', 'program']).withSensor(['IN', 'AL', 'OU'], ea.STATE_SET)],
onEvent: tuya.onEventSetLocalTime,
},
];
```
And when all the devices are added, you replace it and restart the Z2M. Everything is working.

**Reply by Westcott1** [reply]
Unfortunately this re-introduces the temperature 1/10 error.

**Reply by Fnordcorps** [reply]
Has anyone worked out a fix for the temperature problem, this works great and thanks but I cant get it to display the correct temp

**Reply by Westcott1** [reply]
Tolwi's Avatto converter code (above) displays temperature correctly.
Anything based on existing Moes code will have the wrong value.
It seems like Avatto sends the temperature as an integer, all other manufacturers send (temperature with one decimal place *10) as an int, so 21.5 is sent as 215, and converted by dividing by 10.

**Reply by mifrith** [reply]
> Unfortunately this re-introduces the temperature 1/10 error.

I meant from the beginning we insert your code into the converter and add devices, then replace it with the Tolwi code and everything works great.

**Comment by Westcott1** [comment]
Hi Tolwi,
Thanks for doing it properly!
You will have noticed that my attempt was to just get something working.
I  got when it joined -
`zigbee-herdsman-converters:AVATTO: Unrecognized DP #108 with data {"status":0,"transid":2,"dp":108,"datatype":0,"fn":0,"data":{"type":"Buffer","data":[6,0,20,8,0,15,11,30,15,12,30,15,17,0,22,22,0,15,8,0,22,23,0,15]}}`
Thanks again,
Hamish.

**Reply by mifrith** [reply]
Maybe a little late, but still, I found this converter: https://neatho.me/ru/help/termostat-avatto-zwt81/
Everything seems to work, I approached.

**Reply by Westcott1** [reply]
Great, best converter so far!
Thanks for posting

**Reply by tmaniac** [reply]
unfortunately it does not work for me :(

Getting a load of errors upon changing settings etc.

Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Exception while calling fromZigbee converter: tuya.firstDpValue is not a function}
Error Publish 'get' 'system_mode' to '0xb4e3f9fffe7c4ca5' failed: 'Error: Cluster 'hvacThermostat' has no attribute 'SystemMode''
Error Publish 'get' 'system_mode' to '0xb4e3f9fffe7c4ca5' failed: 'Error: Cluster 'hvacThermostat' has no attribute 'SystemMode''

**Reply by mifrith** [reply]
If you are using z2m in docker, you need to uncomment the lines:
const dp = msg.data.dp;
const value = tuya.getDataValue(msg.data.datatype, msg.data.data);
let temperature;
/*meta.logger.warn(`COMMAND #${dp} with data ${JSON.stringify(msg.data)}`);
and comment out:
/*const dpValue = tuya.firstDpValue(msg, meta, 'moes_thermostat');
const dp = dpValue.dp;
const value = tuya.getDataValue(dpValue);
let temperature;
Then everything will work, for some reason in version 1.22.2 the converters are not updated. We are waiting for a new version.

**Comment by Sackjon** [comment]
This now appears to have been included as part of https://www.zigbee2mqtt.io/devices/BHT-002-GCLZB.html#moes-bht-002-gclzb

**Reply by Elkropac** [reply]
Anyone using this device? I was using converter from this discusion (maybe, maybe from some external link), then i switched to moes-bht-002. Now it seems, it's broken. I cannot set program times in "exposes" tab. Also, it seems i cannot  start the "program" preset. Maybe zigbee2mqtt is broken now somehow, there seems to be multiple issues reported on thermostats
